packages/frontend/src/components/staking/components/Validator.js

```diff
@@ -5,47 +5,59 @@ import { useDispatch, useSelector } from 'react-redux';
 import { useNEARAsTokenWithMetadata } from '../../../hooks/fungibleTokensIncludingNEAR';
 import { Mixpanel } from '../../../mixpanel';
 import { redirectTo } from '../../../redux/actions/account';
+import { claimFarmRewards, getValidatorFarmData } from '../../../redux/actions/staking';
 import { selectAccountId } from '../../../redux/slices/account';
-import { actions as tokensActions, selectAllContractMetadata } from '../../../redux/slices/tokens';
+import { selectValidatorsFarmData, selectFarmValidatorAPY } from '../../../redux/slices/staking';
+import { selectTokensFiatValueUSD, selectTokenWhiteList } from '../../../redux/slices/tokenFiatValues';
+import { selectAllContractMetadata } from '../../../redux/slices/tokens';
 import { actionsPending } from '../../../utils/alerts';
-import { PROJECT_VALIDATOR_VERSION, ValidatorVersion } from '../../../utils/constants';
+import { FARMING_VALIDATOR_VERSION } from '../../../utils/constants';
 import FormButton from '../../common/FormButton';
 import SafeTranslate from '../../SafeTranslate';
 import AlertBanner from './AlertBanner';
 import BalanceBox from './BalanceBox';
+import ClaimConfirmModal from './ClaimConfirmModal';
+import { FarmingAPY } from './FarmingAPY';
 import StakeConfirmModal from './StakeConfirmModal';
 import StakingFee from './StakingFee';
 
-const { fetchToken } = tokensActions;
-
-const renderFarmUi = ({ farmList, contractMetadataByContractId, isFarmListLoading }) => {
-    if(isFarmListLoading) {
+const renderFarmUi = ({ farmList, contractMetadataByContractId, openModal, tokenPriceMetadata }) => {
+    if (!farmList.length) {
         // eslint-disable-next-line jsx-a11y/heading-has-content
-        return <h1 className="animated-dots"/>;
+        return <h1 className="animated-dots" />;
     }
 
-    return farmList.map(({ token_id, balance }, i) => {
+    return farmList.map((farm, i) => {
+        const { token_id, balance } = farm;
         const currentTokenContractMetadata = contractMetadataByContractId[token_id];
 
         if (!currentTokenContractMetadata) {
             return;
         }
+        const fiatValueMetadata = tokenPriceMetadata.tokenFiatValues[token_id];
+        const isWhiteListed = tokenPriceMetadata.tokenWhitelist.includes(token_id);
```

Comment:
Is storing tokenWhitelist as a map/set instead of array suitable for this case? Getting data from a set/map will be faster than searching.

Reply:
I'm not sure that this has any sense. As we receive just an array of token ids (`["wrap.near", "aurora", ...]`).